internal/controller/customreplicaset_controller.go

```diff
@@ -336,215 +359,223 @@ func (r *CustomReplicaSetReconciler) findChildPods(ctx context.Context, req ctrl
+				return err
+			}
 		}
-		fmt.Println("Created Pod")
-		totalPods++
-	}
-	return nil
-}
 
-func (r *CustomReplicaSetReconciler) deletePods(ctx context.Context, totalPods, stdPods, upgPods int, cr *customreplicasetv1.CustomReplicaSet, childPods corev1.PodList, log logr.Logger) error {
-	podsToDelete := totalPods - int(cr.Spec.Replicas)
-	upgPodsToDelete := upgPods - int(cr.Spec.Partition)
-	stdPodsToDelete := podsToDelete - upgPodsToDelete
-
-	for _, pod := range childPods.Items {
-		if pod.Status.Phase == corev1.PodRunning || pod.Status.Phase == corev1.PodPending {
-			if strings.Contains(pod.ObjectMeta.Name, "upgraded") {
-				if upgPodsToDelete > 0 {
-					upgPodsToDelete--
-				} else {
-					continue
-				}
-			} else {
-				if stdPodsToDelete > 0 {
-					stdPodsToDelete--
-				} else {
-					continue
-				}
+		if oldPodsToCreate > 0 {
+			oldRevision, err := r.getControllerRevisionExcludingRevision(ctx, crs, latestRevision.Revision)
+			if err != nil {
+				fmt.Println("Failed to get a controller revision excluding the latest one")
+				return err
 			}
-
-			if err := r.Delete(ctx, &pod); err != nil {
-				log.Error(err, "Unable to delete pod")
+			if err := r.createPods(ctx, crs, oldPodsToCreate, podsPerRevision, oldRevision); err != nil {
+				fmt.Println("Failed to create pods", err)
+				return err
+			}
+		}
+	} else if totalAvailPods > int(crs.Spec.Replicas) {
+		podsToDelete := totalAvailPods - int(crs.Spec.Replicas)
+		newPodsToDelete := len(latestPods) - int(crs.Spec.Partition)
+		oldPodsToDelete := podsToDelete - newPodsToDelete
+		if newPodsToDelete > 0 {
+			if err := r.deletePods(ctx, crs, newPodsToDelete, podsPerRevision, childPods, true); err != nil {
+				fmt.Println("Failed to delete pods", err)
 				return err
 			}
-			fmt.Println("Delete Pod")
-			totalPods--
-			if totalPods == int(cr.Spec.Replicas) {
-				break
+		}
+		if oldPodsToDelete > 0 {
+			if err := r.deletePods(ctx, crs, oldPodsToDelete, podsPerRevision, childPods, false); err != nil {
+				fmt.Println("Failed to delete pods", err)
+				return err
 			}
 		}
+	} else {
+		if err := r.upgradeOrDowngradePods(ctx, crs, totalAvailPods, podsPerRevision, latestRevision, childPods, len(latestPods)); err != nil {
+			fmt.Println("Failed to upgrade or downgrade pods", err)
+			return err
+		}
 	}
 	return nil
 }
 
-func (r *CustomReplicaSetReconciler) upgradeOrDowngradePods(ctx context.Context, stdPods, upgPods int, cr *customreplicasetv1.CustomReplicaSet, childPods corev1.PodList, log logr.Logger) error {
-	// Check if we need to upgrade or downgrade any pods
-	podsToUpgrade := int(cr.Spec.Partition) - upgPods
-	if podsToUpgrade > 0 {
-		if err := r.upgradePods(ctx, podsToUpgrade, cr, childPods, log); err != nil {
-			log.Error(err, "Unable to upgrade pod")
+func (r *CustomReplicaSetReconciler) createPods(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, podsToCreate int, podsPerRevision map[int][]*corev1.Pod, revision *v1.ControllerRevision) error {
```

Comment:
podsPerRevision is not needed in the method. It can be removed from the parameter list.

Reply:
Right thanks removed